kvbc/src/ReplicaImp.cpp

```diff
@@ -263,39 +255,38 @@ ReplicaImp::~ReplicaImp() {
   }
 }
 
-Status ReplicaImp::addBlockInternal(const SetOfKeyValuePairs &updates, BlockId &outBlockId) {
-  outBlockId = m_bcDbAdapter->addBlock(updates);
-
-  return Status::OK();
-}
-
-Status ReplicaImp::getInternal(BlockId readVersion, const Key &key, Sliver &outValue, BlockId &outBlock) const {
-  const auto clear = [&outValue, &outBlock]() {
-    outValue = Sliver{};
-    outBlock = 0;
-  };
-
-  try {
-    std::tie(outValue, outBlock) = m_bcDbAdapter->getValue(key, readVersion);
-  } catch (const NotFoundException &) {
-    clear();
-  } catch (const std::exception &e) {
-    clear();
-    return Status::GeneralError(std::string{"getInternal() failed to get value due to a DBAdapter error: "} + e.what());
-  } catch (...) {
-    clear();
-    return Status::GeneralError("getInternal() failed to get value due to an unknown DBAdapter error");
-  }
-
-  return Status::OK();
-}
-
 /*
  * This method can't return false by current insertBlockInternal impl.
  * It is used only by State Transfer to synchronize state between replicas.
  */
-bool ReplicaImp::putBlock(const uint64_t blockId, const char *block_data, const uint32_t blockSize) {
-  Sliver block = Sliver::copy(block_data, blockSize);
+bool ReplicaImp::putBlock(const uint64_t blockId, const char *blockData, const uint32_t blockSize) {
+  if (replicaConfig_.isReadOnly) {
+    return putBlockToObjectStore(blockId, blockData, blockSize);
+  }
+
+  auto view = std::string_view{blockData, blockSize};
+  const auto rawBlock = categorization::RawBlock::deserialize(view);
+  if (m_kvBlockchain->hasBlock(blockId)) {
+    const auto existingRawBlock = m_kvBlockchain->getRawBlock(blockId);
+    if (rawBlock != existingRawBlock) {
+      LOG_ERROR(logger,
+                "found existing (and different) block ID[" << blockId << "] when receiving from state transfer");
+
+      // TODO consider assert?
+      m_kvBlockchain->deleteBlock(blockId);
```

Comment:
Are we sure we want to delete the existing block? Is it better to just crash the node?

I think this can happen in 2 ways:
 * More than F replicas are malicious and therefore the received block is bad.
 * The disk state is corrupt.

In both cases it may be better to preserve the evidence.

Reply:
I just wanted to preserve the existing behaviour in this PR. I agree with you, though.

Reply:
Ah, I didn't realize it was existing behavior.